XiEditor/XiTextPlane/Atlas.swift

```diff
@@ -0,0 +1,200 @@
+        self.ctFont = ctFont
+    }
+}
+
+func fontMapKey(glyph: CGGlyph, flags: UInt32) -> UInt32 {
+    return flags | UInt32(glyph)
+}
+
+typealias FontRef = Int
+
+/// Texture atlas containing glyph cache.
+
+class Atlas {
+    var textureId: GLuint = 0
+    let width = 512
+    let height = 512
+    let uvScale: GLfloat
+    
+    // key is rounded-up height, value is x, y coords of next alloc
+    var strips: [Int: (Int, Int)] = [:]
+    var nextStrip = 0
+
+    var fontCache = FontCache()
+    var fontRefs: [CTFont: FontRef] = [:]
+    
+    init() {
+        glGenTextures(1, &textureId)
+        glBindTexture(GLenum(GL_TEXTURE_2D), textureId)
+        glPixelStorei(GLenum(GL_UNPACK_ALIGNMENT), 1)
+        glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_WRAP_S), GL_CLAMP_TO_EDGE)
+        glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_WRAP_T), GL_CLAMP_TO_EDGE)
+        glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_MAG_FILTER), GL_NEAREST)
+        glTexParameteri(GLenum(GL_TEXTURE_2D), GLenum(GL_TEXTURE_MIN_FILTER), GL_NEAREST)
+        var data = [UInt8](repeating: 255, count: width * height * 4)
+
+        glTexImage2D(GLenum(GL_TEXTURE_2D), 0, GL_RGBA, GLsizei(width), GLsizei(height), 0, GLenum(GL_BGRA), GLenum(GL_UNSIGNED_BYTE), &data)
+        uvScale = 1.0 / Float(width)
+    }
+    
+    // Round a value up to the next power-of-2 multiple of a number at most fracBits long
+    func roundUp(_ y: Int, fracBits: Int) -> Int {
+        if y < 1 << fracBits { return y }
+        let mask = 1 << (Int.bitWidth - (y - 1).leadingZeroBitCount - fracBits)
+        return y + (-y & (mask - 1))
+    }
+    
+    // Returns x, y coords on successful alloc
+    func allocRect(w: Int, h: Int) -> (Int, Int)? {
+        if w > width || h > height {
+            // request cannot be satisfied
+            return nil
+        }
+        let roundHeight = roundUp(h, fracBits: 2)
+        var coords = strips[roundHeight]
+        if coords == nil || coords!.0 + w > width {
+            if nextStrip + roundHeight > height {
+                // no more free space in this atlas
+                return nil
+            }
+            coords = (0, nextStrip)
+            nextStrip += roundHeight
+        }
+        strips[roundHeight] = (coords!.0 + w, coords!.1)
+        return coords
+    }
+
+    func getGlyph(fr: FontRef, glyph: CGGlyph, flags: UInt32, scale: CGFloat) -> CachedGlyph? {
+        // Some discussion about this approach: the correctness condition is that we never
+        // use the same key for two different scale factors. A more robust approach would
+        // be to generate a unique flag value for each scale factor, but this will work in
+        // the common case of two monitors, one low-dpi (1.0) and the other retina (2.0).
+        // It will also work on devices with a single scaling factor, no matter the value.
+        let flagsForKey = flags | (scale > 1.0 ? FLAG_HI_DPI : 0)
```

Comment:
This is different. The fake italic flag is part of the request, it affects the result. The hidpi flag is there solely to avoid cache poisoning; the actual request is the scale parameter.

I agree it's not very clean. Perhaps a cleaner way would be to have a registry of dpi values in some shared state (Atlas is probably the one that makes the most sense), a method to look up a dpi scale and return a flag value, then the flag is the source of truth. We'd bound the number of distinct dpi values at something reasonable (I believe a value of 2 will cover 99.999% of systems, but we can use 4 bits or something). It would be more correct in the cases we discussed, and arguably cleaner. I'm willing to make the change if you agree it's a serious improvement.

I'm pretty sure that TextLine is dpi-independent; the scale value is applied at drawGlyphInstance time. This design allows a TextLine to be reused when moving from one screen to another (and avoids the need to do explicit work in response to a dpi change notification). Perhaps that's another reason to keep them separate.

I would totally use a tuple for the atlas lookup if it worked, but it doesn't. Given that the key is a somewhat complicated mixture of glyph index and various bits set for various reasons, I personally feel that munging bits at the low level captures the intent more clearly than mixing OptionSet with other manipulations to get at the tuple-ish fields.

Reply:
Good rationale, thanks. I think the thing that makes this most confusing is that DPI gets set by the layer, but all the renderer calls are in the view. If the view also made the startDraw/endDraw calls, this would be clearer; at the very least it is more likely that a reader of the drawing code would understand that we were explicitly taking `contentScale` into account on each draw call.

I think that your proposed solution is a nice one, but I also don't think it's crucial. It is unclear to me that this will ever be an actual problem. So maybe put it in the cleanup issue?

If we're defensive, we could also check if the contentScale is not in [1.0, 2.0] and log it/abort (probably not abort).